### ISSUE TITLE
CSS :active doesn't work on Firefox

### ISSUE THREAD

**Issue by rogerl50** [ArnoldSmith86/virtualtabletop#1169]
A button with the following CSS:
```JSON
"css": {
    "default": {
      "background-color": "green"
    },
    ":hover": {
      "background-color": "blue"
    },
    ":active": {
      "background-color": "red"
    }
  }
```
will work properly on Chrome and Edge. But on Firefox, while the button turns blue on hover, it does not turn red when pressed. On the other hand, similar code at https://www.w3schools.com/cssref/tryit.asp?filename=trycss_sel_link_more1 and at https://jsfiddle.net/rxl50/bLvrnjfo/ work properly on Firefox.

**Comment by robartsd** [comment]
I wonder if our mouse handling code is interfering with native :active treatment in firefox in some way that does not get treated the same way by chromium brownsers.

**Comment by 96LawDawg** [comment]
> I wonder if our mouse handling code is interfering with native :active treatment in firefox in some way that does not get treated the same way by chromium brownsers.

That might be a good guess.  I just tried a couple of other things using :active in Firefox.  Nothing triggered.  It looks like it just isn't working in FF the same way it does in the other browsers.

**Comment by ArnoldSmith86** [comment]
Yeah, that would be my guess as well. We basically tell the browser for every possible mouse interaction "forget you wanted to do and do what we tell you to".

My guess is that you only want `:active` because you want to style buttons differently while they being clicked. My suggestion would be that we manually add a `.clicked` class to the widget instead of trying to repair `:active`. I would make it client-only (so only the player clicking the button would see the change) but we could also send that information to other players.

**Comment by rogerl50** [comment]
> My guess is that you only want `:active` because you want to style buttons differently while they being clicked. 
Yes, exactly.
> My suggestion would be that we manually add a `.clicked` class to the widget instead of trying to repair `:active`. I would make it client-only (so only the player clicking the button would see the change) but we could also send that information to other players.
I would send it to other players as well. The ability to have something that looks like a real button is pretty nice and would really spiff up the UI IMO.

**Comment by 96LawDawg** [comment]
> I would make it client-only (so only the player clicking the button would see the change) but we could also send that information to other players.

Sending it to other players would give the option for room designers to give an indication to others that a button has actually been pressed.  That would be a nice addition in some games.

**Comment by rogerl50** [comment]
I just figured out that the `transform` css property also doesn't get translated properly, since `transform` is used to move the widget coordinates, and user-specified transforms are not appended (if that's even possible). I wonder what other restrictions there are on what css you can specify in VTT.

**Comment by 96LawDawg** [comment]
> I just figured out that the `transform` css property also doesn't get translated properly, since `transform` is used to move the widget coordinates, and user-specified transforms are not appended (if that's even possible). I wonder what other restrictions there are on what css you can specify in VTT.

You can sort of make `transform` work if you use !important.  @robartsd gave an example for me recently at https://discord.com/channels/770758631146782780/793626590848352306/955921962906038422.

**Comment by robartsd** [comment]
Yes, you'd want to recreate the native transform with `transform: translate(${PROPERTY x}px ${PROPERTY y}px) rotate(${PROPERTY rotation}deg) scale(${PROPERTY scale)` (changing any part of the transform you actually intend to change of course).

I suggested a potential `clicking` property in #1145.

**Comment by ArnoldSmith86** [comment]
I'm aware that everybody would prefer for every player to see the new `.clicked` state.

The problem is that it's basically impossible to know if the state does look different or not (so the engine has to send `clicking` for every click on any widget ever - basically doubling the deltas that have to be sent). And on top of that, the new delta has to be sent on `mousedown`, at which point the engine can't even know if the player is clicking or dragging the widget (except that it could look for `clickable: false`) so it not only has to send new deltas when clicking widgets but also when dragging widgets.

It's probably not a big deal with all the cursor events we send anyway but I just want to make sure that everyone understands the repercussions of sending the `clicking` to other players. It's not "I have this reset button over here, so the engine has to send one more delta", it is "any widget could have this, so the engine has to send double the deltas in total".

**Comment by 96LawDawg** [comment]
> The problem is that it's basically impossible to know if the state does look different or not (so the engine has to send `clicking` for every click on any widget ever - basically doubling the deltas that have to be sent). 

Pardon my ignorance, but wouldn't you only have to send double deltas for widgets that specifically invoke the `.clicked` state/class?  If I'm wrong, no need to explain why I'm wrong.  Just making sure that I try to follow what is going on.

**Comment by ArnoldSmith86** [comment]
> Pardon my ignorance, but wouldn't you only have to send double deltas for widgets that specifically invoke the `.clicked` state/class?

With the freedom of `css` you could define that CSS in any other widget. It might be a good idea to enforce it though so that it only works for widgets that use `.clicked` in _their own_ `css`.
